File: 2020/knowit/18/18.js
"use strict";

const fs = require("fs");

const isPalindrome = word => {
	const l = word.length;
	const half = Math.floor(l / 2);
	for (let i = 0; i < half; i++) {
		if (word[i] !== word[l - i - 1]) {
			return false;
		}
	}
	return true;
}

const isAlmostPalindrome = word => {
	if (word.length < 3) {
		return false;
	}

	if (isPalindrome(word)) {
		return false;
	}

	const l = word.length;
	const half = Math.floor(l / 2);
	let i = 0;
	while (i < half) {
		if (word[i] === word[l-i-1]) {
			i++;
		} else if (word[i] === word[l-i-2] && word[i+1] === word[l-i-1]) {
			i += 2;
		} else {
			return false;
		}
	}

	return true;
};

const almostPalindromes =
	fs
		.readFileSync("wordlist.txt", "utf-8")
		.split("\n")
	//["gnisning", "kauka", "baluba", "tarotkorta"]
		.filter(isAlmostPalindrome)
		.length;

console.log(almostPalindromes);

module.exports = {
	isAlmostPalindrome,
	isPalindrome
};

File: 2020/knowit/18/18.hs
import System.IO (readFile)

isAlmostPalindrome :: String -> Bool
isAlmostPalindrome word
    | l < 3         = False
    | word == rWord = False
    | otherwise     = isAlmostPalindrome' word rWord
    where l = length word
          rWord = reverse word
          isAlmostPalindrome' :: String -> String -> Bool
          isAlmostPalindrome' [] [] = True
          isAlmostPalindrome' (a:[]) (a':[]) = a == a'
          isAlmostPalindrome' (a:b:rest) (a':b':rest')
              | a == a'            = isAlmostPalindrome' (b:rest) (b':rest')
              | a == b' && a' == b = isAlmostPalindrome' rest rest'
              | otherwise          = False

main =
    print
    . length
    . filter isAlmostPalindrome
    . lines
    =<< readFile "wordlist.txt"
